common/namespace/registry.go

```diff
@@ -483,6 +486,151 @@ func (r *registry) getNamespaceByIDLocked(id ID) (*Namespace, error) {
 	return nil, serviceerror.NewNamespaceNotFound(id.String())
 }
 
+// getOrPutNamespace retrieves the information from the cache if it exists or reads through
+// to the persistence layer and updates caches if it doesn't
+func (r *registry) getOrPutNamespace(name Name) (*Namespace, error) {
+	// check main caches
+	cacheHit, cacheErr := r.getNamespace(name)
+	if cacheErr == nil {
+		return cacheHit, nil
+	}
+
+	r.readthroughLock.Lock()
+	defer r.readthroughLock.Unlock()
+
+	// check caches again in case there was an update while waiting
+	cacheHit, cacheErr = r.getNamespace(name)
+	if cacheErr == nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough cache
+	if cachedNotFound := r.readthroughNotFoundCache.Get(name.String()); cachedNotFound != nil {
+		return nil, serviceerror.NewNamespaceNotFound(name.String())
+	}
+
+	// readthrough to persistence layer and update readthrough cache if not found
+	ns, err := r.getNamespaceByNamePersistence(name)
+	if err != nil {
+		return nil, err
+	}
+
+	// update main caches if found
+	r.updateCachesSingleNamespace(ns)
+
+	return ns, nil
+}
+
+// getOrPutNamespaceByID retrieves the information from the cache if it exists or reads through
+// to the persistence layer and updates caches if it doesn't
+func (r *registry) getOrPutNamespaceByID(id ID) (*Namespace, error) {
+	// check main caches
+	cacheHit, cacheErr := r.getNamespaceByID(id)
+	if cacheErr == nil {
+		return cacheHit, nil
+	}
+
+	r.readthroughLock.Lock()
+	defer r.readthroughLock.Unlock()
+
+	// check caches again in case there was an update while waiting
+	cacheHit, cacheErr = r.getNamespaceByID(id)
+	if cacheErr == nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough cache
+	if cachedNotFound := r.readthroughNotFoundCache.Get(id.String()); cachedNotFound != nil {
+		return nil, serviceerror.NewNamespaceNotFound(id.String())
+	}
+
+	// readthrough to persistence layer and update readthrough cache if not found
+	ns, err := r.getNamespaceByIDPersistence(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// update main caches if found
+	r.updateCachesSingleNamespace(ns)
+
+	return ns, nil
+}
+
+func (r *registry) updateCachesSingleNamespace(ns *Namespace) {
+	r.cacheLock.Lock()
+	defer r.cacheLock.Unlock()
+
+	if curEntry, ok := r.cacheByID.Get(ns.ID()).(*Namespace); ok {
+		if curEntry.NotificationVersion() >= ns.NotificationVersion() {
+			// More up to date version already put in cache by refresh
+			return
+		}
+	}
+
+	oldNS := r.updateIDToNamespaceCache(r.cacheByID, ns.ID(), ns)
+	r.cacheNameToID.Put(ns.Name(), ns.ID())
+	if namespaceStateChanged(oldNS, ns) {
+		r.stateChangedDuringReadthrough = append(r.stateChangedDuringReadthrough, ns)
```

Comment:
this means callback can be delayed for 10s for new namespaces.
The only problem I can think of is that, say user runs a Schedule immediately after registering a ns, there will be a 10s delay, which should be ok?

I don't have a strong opinion here but seems like we are combining two things together:
- load or refresh namespaces
- trigger callback

I kinda wonder if it's easy to split them into different places, i.e.:
- one goroutine only do refresh, store namespace that needs callback somewhere. trigger callback (send a signal through channel or something like that)
- do the same thing for the readthrough path
- after callback goroutine get the signal, copy those namespaces that needs callback out and then run the callbacks.

Current approach also looks fine to me.

Reply:
@dnr Did you have any thoughts on approach here? Using a separate goroutine for running callbacks makes sense to me from a separation of responsibilities perspective, but it is slightly more complicated. I'm ambivalent; either seems fine to me.

Reply:
Second: I like the idea of a goroutine just for callbacks, but I also want to combine the refresh and readthrough path as much as possible and clean up that part. What do you think of this?
```
// nss are one or more namespaces just read from persistence.
// complete is true if they were from ListNamespaces and so everything
// not in nss can be assumed to be deleted.
func mergeNamespacesIntoCache(nss []*Namespace, complete bool) {
  // lock+defer cache lock
  // merge nss into cache, adding callbacks to deferred list if different
  // also, if complete is true, remove deleted from cache
}

func getNamespace(name) {
  // check cache (under cache lock)
  // lock+defer persistence lock
  // check negative cache
  readOneNSFromPersistence()
  mergeNamespacesIntoCache([]{ns}, false)
}

func refreshOnce() {
  // lock+defer persistence lock
  nss := ListNamespaces()
  mergeNamespacesIntoCache(nss, true)
}

func refreshLoop() {
  for {
    refreshOnce()
    callDeferredCallbacks()
  }
}
```

That can also be combined with moving callbacks to another goroutine: I think we could just have a `chan []struct{ns *Namespace, deletedFromDb bool}` and just call them in a loop? If the buffer fills up we might block getNamespace which is unfortunate. Could also do a hybrid, try to send on a channel, if would block then add to deferred list and call in refresh loop.

I don't think we need to "copy" anything, Namespace is immutable, we can just pass pointers to them